common/meta/api/src/schema_api_impl.rs

```diff
@@ -196,6 +263,101 @@ impl<KV: KVApi> SchemaApi for KV {
         }
     }
 
+    async fn undrop_database(
+        &self,
+        req: UndropDatabaseReq,
+    ) -> Result<UndropDatabaseReply, MetaError> {
+        let name_key = &req.name_ident;
+
+        loop {
+            let res =
+                get_db_or_err(self, name_key, format!("undrop_database: {}", &name_key)).await;
+
+            if res.is_ok() {
+                return Err(MetaError::AppError(AppError::UndropDbAlreadyExists(
+                    UndropDbAlreadyExists::new(&name_key.db_name),
+                )));
+            }
+
+            // get db id list from _fd_db_id_list/<tenant>/<db_name>
+            let dbid_idlist = DbIdListKey {
+                tenant: name_key.tenant.clone(),
+                db_name: name_key.db_name.clone(),
+            };
+            let (db_id_list_seq, db_id_list_opt): (_, Option<DbIdList>) =
+                get_struct_value(self, &dbid_idlist).await?;
+
+            let mut db_id_list = if db_id_list_seq == 0 {
+                return Err(MetaError::AppError(AppError::UndropDbHasNoHistory(
+                    UndropDbHasNoHistory::new(&name_key.db_name),
+                )));
+            } else {
+                match db_id_list_opt {
+                    Some(list) => list,
+                    None => {
+                        return Err(MetaError::AppError(AppError::UndropDbHasNoHistory(
+                            UndropDbHasNoHistory::new(&name_key.db_name),
+                        )))
+                    }
+                }
```

Comment:
Reading the `db_id` and reading the `db_id_list` is not guaranteed to happen in a  **snapshot**.
Thus there is a chance `db_id_list` is inconsistent and in such a case it should retry in next loop.